# Required packages:
# - Trim Galore
# - Fastqc
# - Bowtie2
# - Bismark
# - Samtools

R1_fastq="Raw fastq file with read 1"
R2_fastq="Raw fastq file with read 2"
samp_name="Sample name"


# Trimming adapter sequences with Trim Galore from the raw fastq files
trim_galore --paired $R1_fastq $R2_fastq

trimmed_R1="$(echo $R1_fastq | cut -d '/' -f8)"
trimmed_R2="$(echo $R2_fastq | cut -d '/' -f8)"
prefix_R1="$(echo $trimmed_R1 | cut -d '.' -f1)"
prefix_R2="$(echo $trimmed_R2 | cut -d '.' -f1)"

# Performing quality control check of the sequencing data
fastqc -t 10 ${prefix_R1}*.fq.gz
fastqc -t 10 ${prefix_R2}*.fq.gz

# Alignment of reads against mouse (mm10) reference genome using Bismark. 
# 'mm10_genome_dir' is a directory generated by the bismark_genome_preparation function from Bismark, which uses a fasta file with the mm10 genome
# and prepares the genome for bisulfite alignments.
bismark --bowtie2 --bam mm10_genome_dir -1 ${prefix_R1}*.fq.gz -2 ${prefix_R2}*.fq.gz

# Sorting the BAM file by read name
samtools sort -n ${samp_name}_nsorted.bam ${samp_name}_bismark_bt2_pe.bam

# Removing duplicate reads
deduplicate_bismark -p -o ${samp_name} --bam ${samp_name}_nsorted.bam

# Generating M-bias plot
# M-bias plots are used to detect CpG methylation bias across the length of the read. The M-bias plot should inform how many base pairs should be trimmed
# from the 5' and 3' in reads 1 and 2 so they are not included in the downstream anlysis. Such number of base pairs with bias will specified in the informME pipeline.
bismark_methylation_extractor -p --report --mbias_only --output ${samp_name}*deduplicated.bam

# Sorting the BAM file by coordinates
# These BAM files, sorted by coordinates, will be the input for the informME pipeline.
samtools sort -o ${samp_name}_coord_sorted.bam ${samp_name}*deduplicated.bam

# Index sorted BAM file
samtools index ${samp_name}_coord_sorted.bam


# Further processing for identification of block differentially methylated regions with dmrseq:
# Sorting the BAM file by read name
samtools sort -n -o ${samp_name}_name_sorted.bam ${samp_name}_coord_sorted.bam

# Generating BEDmethyl files (Bismark CpG reports and coverage files) with methylated and unmethylated counts per CpG from BAM files. 
# The Bismark CpG reports will be the input to construct the bsseq object in R for downstream analysis with dmrseq and detection of block DMRs.
# The Bismark coverage files will be the input for the cell type deconvolution analysisis with the MeDeCom pipeline.
bismark_methylation_extractor -p --report --bedGraph --gzip --comprehensive --merge_non_CpG --counts --cytosine_report --ignore 5 --ignore_r2 20 --genome_folder mm10_genome ${samp_name}_name_sorted.bam


