# Required packages:
# - Trim Galore
# - Fastqc
# - Bowtie2
# - Bismark
# - Samtools

R1_fastq="Raw fastq file with read 1"
R2_fastq="Raw fastq file with read 2"
samp_name="Sample name"


# Trimming adapter sequences with Trim Galore
trim_galore --paired $R1_fastq $R2_fastq

trimmed_R1="$(echo $R1_fastq | cut -d '/' -f8)"
trimmed_R2="$(echo $R2_fastq | cut -d '/' -f8)"
prefix_R1="$(echo $trimmed_R1 | cut -d '.' -f1)"
prefix_R2="$(echo $trimmed_R2 | cut -d '.' -f1)"

# Performing quality control check of the sequencing data
fastqc -t 10 ${prefix_R1}*.fq.gz
fastqc -t 10 ${prefix_R2}*.fq.gz

# Alignment of reads against mouse (mm10) reference genome using Bismark. 
# 'mm10_genome_dir' is a directory generated by the bismark_genome_preparation function from Bismark, which uses a fasta file with the mm10 genome and prepares the genome for bisulfite alignments.
bismark --bowtie2 --bam mm10_genome_dir -1 ${prefix_R1}*.fq.gz -2 ${prefix_R2}*.fq.gz

# Sorting the BAM file by read name
samtools sort -n ${samp_name}_nsorted.bam ${samp_name}_bismark_bt2_pe.bam

# Removing duplicate reads
deduplicate_bismark -p -o ${samp_name} --bam ${samp_name}_nsorted.bam

# Generating M-bias plot
# M-bias plots are used to detect CpG methylation bias across the length of the read. The M-bias plot should inform how many base pairs should be trimmed from the 5' and 3' in reads 1 and 2 so they are not included in the downstreama anlysis.
bismark_methylation_extractor -p --report --mbias_only --output ${samp_name}*deduplicated.bam






# For informME/Mauro parsing:
echo "Running coordinate sorting... on $outdir/${samp_name}_deduplicated.bam"
samtools sort --threads 12 -o $outdir/${samp_name}_coord_sorted.bam $outdir/${samp_name}*deduplicated.bam

echo "Running indexing of BAM file sorted by coordinates"
samtools index -@ 16 ${samp_name}_coord_sorted.bam


